docs/ds20.md

```diff
@@ -0,0 +1,111 @@
+We do need to use the `BOS` *“Binary Object Store”* descriptor, which is only supported for devices using USB 2.1 and newer, although that’s probably the vast majority of devices now.
+Usually the BOS descriptor is used for Wireless USB details, USB 2.0 extensions, SuperSpeed USB connection details and a Container ID.
+
+Usefully, the BOS descriptor does give us is the ability to use the platform capability descriptor, which is `bDevCapabilityType=0x05`. For UUID `D8DD60DF-4589-4CC7-9CD2-659D9E648A9F`, this is identified as a structured blob of data Windows usually uses to put workarounds like the suspend mode of the device and that kind of thing.
+
+We could certainly create a new `bDevCapabilityType` which would allow us to store a binary blob (e.g. `Plugin=foobarbaz\nFlags=QuirkValueHere\n`) but that would be out-of-specification and difficult for vendors to implement.
+
+## Proposal
+
+Devices create a new platform device capability BOS descriptor with `bDevCapabilityType=0x05` and a fwupd-specific UUID, specifically `010aec63-f574-52cd-9dda-2852550d94f0`.
+Using this custom UUID will ensure that Microsoft Windows does not try to parse the capability descriptor as a *Microsoft OS 2.0 descriptor set*.
+
+## Implementation
+
+Create a BOS DS20 descriptor such as:
+
+    1C 10 05 00 63 ec 0a 01 74 f5 cd 52 9d da 28 52 55 0d 94 f0 05 08 01 00 20 00 2a 00
+    |  |  |  |  \---------------------------------------------/ \--------/  \---/ |  |
+    |  |  |  |                ^-PlatformCapabilityUUID            |   wLength-^   |  |
+    |  |  |  ^-bReserved                                 dwVersion^   bVendorCode-^  |
+    |  |  ^-bDevCapability                                                bAltEnumCmd^
+    |  ^-bDescriptorType
+    ^-bLength
+
+The `dwVersion` encoded here is fwupd `1.8.5`, which is also the first version that supports BOS DS20 descriptors.
+A `dwVersion` of 0x0 will match any fwupd version, but should probably not be used unless required.
+
+The BOS descriptors are sorted by the requester and can appear in any order.
+The descriptor with the highest `dwVersion` that is not newer than the current running daemon version is used.
+This allows devices to set different quirks depending on the fwupd version, although in practice this should not be required.
+The `bAltEnumCmd` parameter **must** be zero.
+The `PlatformCapabilityUUID` **must** be `010aec63-f574-52cd-9dda-2852550d94f0`.
+
+Then allow the device to reply to a USB control request with the following parameters:
+
+    transfer-direction: device-to-host
+    request-type: vendor
+    recipient: device
+    bRequest: {value of bVendorCode in the BOS descriptor)
+    wValue: 0x00
+    wIndex: 0x07
+    wLength: {value of wLength in the BOS descriptor)
+
+The device should return something like:
+
+    50 6c 75 67 69 6e 3d 64 66 75 0a 49 63 6f 6e 3d 63 6f 6d 70 75 74 65 72 0a 00 00 00 00 00 00 00
+
+...which is the UTF-8 quirk data, e.g.
+
+    Plugin=dfu
+    Icon=computer
+
```

Comment:
> fwupdtool get-fwupd-bos-d20

What about if we included it in the `->to_string()` output? i.e. it would display in the --verbose output of `fwupdtool get-devices`.

Reply:
Here's a thought; we move the BOS descriptor parsing to a new FuUsbDeviceFwPlatform class that subclasses FuUsbDevicePlatform (that subclasses FuFirmware), and the FuUsbDeviceMsPlatform for the MS parts. We add all of them to a GPtrArry in FuUsbDevice->priv->platform_descriptors, and then call ->to_string() on them when dumping the device. This is something I could have used last night for the fuzzing experiment....

I think that breaks up that massive file a bit too. The other way would be to create a FuUsbDevice child (of the original FuUsbDevice) that parses the file on write_firmware() or something, but that's going to make incorporate break, and the debug output look weird -- and it's weird to parse in a fake write.

Reply:
Wow cool! That sounds like a workable flow.

I would still suggest not to use dd though for that last step however. For development purposes don't you want to be able to populate the quirk file with a header to indicate the GUID?

So if you leave a contrib script that "converts" the quirk file it can strip the header, or even take an argument of the GUID in the header to pull out a "single" device from a quirk file...

Reply:
I just wrote `contrib/generate-ds20.py fw-ds20.quirk --bufsz 32` which does all that you say :)